cmd/gke-certificates-controller/app/gke_certificates_controller.go

```diff
@@ -94,9 +105,64 @@ func Run(s *GKECertificatesController) error {
 		signer.handle,
 	)
 
-	sharedInformers.Start(nil)
-	// controller.Run calls block forever.
-	go approveController.Run(5, nil)
-	go signController.Run(5, nil)
-	select {} // block forever
+	run := func(stopCh <-chan struct{}) {
+		sharedInformers.Start(stopCh)
+		// controller.Run calls block forever.
+		go approveController.Run(5, stopCh)
+		go signController.Run(5, stopCh)
+		<-stopCh
+	}
+
+	if s.LeaderElectionConfig.LeaderElect {
+		leaderElectionClient, err := clientset.NewForConfig(restclient.AddUserAgent(kubeconfig, "leader-election"))
+		if err != nil {
+			return err
+		}
+		leaderElectionConfig, err := makeLeaderElectionConfig(s.LeaderElectionConfig, leaderElectionClient, recorder)
+		if err != nil {
+			return err
+		}
+		leaderElectionConfig.Callbacks = leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				utilruntime.HandleError(fmt.Errorf("lost master"))
```

Comment:
Does this stop the approver/signer goroutines?
What's the difference between `utilruntime.HandleError` and `glog.Error`?

Reply:
Is this guaranteed to be fatal?

Reply:
I think it shouldn't be fatal. It should stop the controller loops and restart on re-election.

Reply:
General guidance is that losing leadership should result in process exit. It's extremely difficult to guarantee that an application hasn't lost track of goroutines that, in error, may continue to modify state after the leader transition. https://github.com/kubernetes/kubernetes/pull/16830#issuecomment-158786529

Reply:
I was just doing the same kube-scheduler does, but I'm not sure whether it is guaranteed to be fatal. I changed this to glog.Fatalf.

Reply:
Can you make it `glog.Exit("lost leader election, exiting")`

Reply:
Done :)